Epic 8438 : Assay Data Tutorials

labkey-bpatel opened the pull request:
#### Rationale
Allow adding of assay tutorials to help users understand the assay data.
Spec [here](https://docs.google.com/document/d/1MtSvDRxyOHmpg07XQZdWB4Etbwza0nwhx80s3ZMdMi4/edit#heading=h.s8qmk21y0t38).

#### Changes
* Schema changes to add assay tutorials links and downloadable documents
* ETL updates
* Assay tutorial view, model, and store changes
* Show tutorial in a popup modal
* Embed tutorial display name in a thumbnail

labkey-bpatel committed:
Add tables to handle assay tutorials, update ETLs, add sample data

labkey-bpatel committed:
Update store and model for assay tutorials

labkey-bpatel committed:
Create Assay tutorial view, update to assay model and store, handle doc validation, create module property for assay tutorial documents.

labkey-bpatel committed:
Updates to show tutorial video in a popup modal

labkey-bpatel committed:
Embed text in thumbnail image, update styling on downloadable tutorials

labkey-bpatel committed:
Sample data update

labkey-bpatel committed:
Auto test for assay tutorial section

XingY commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
+
+            if (!errors.hasErrors() && !StringUtils.isBlank(filename))
+            {
+                if (isFileValid(errors, response, filename, basePath))
```
The function name is very confusing, maybe rename to downloadFile?

labkey-bpatel replied:
i agree, now renamed

XingY commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
+        @Override
+        public void validate(StudyDocumentForm form, BindException errors)
+        {
+            if (null == form.getDocumentId())
```
It doesn't seem to me documentId is used by this action?

labkey-bpatel replied:
correct, now removed

XingY commented on webapp/Connector/src/app/view/module/AssayTutorial.js:
```diff
@@ -0,0 +1,103 @@
+
+    hasContent : function() {
+        var d = this.getListData();
+        return d["assayTutorialDocuments"].length > 0 || d["assayTutorialLinks"].length > 0;
```
can use d["assayTutorialDocuments"]?.length > 0 || d["assayTutorialLinks"]?.length > 0 to check for null

labkey-bpatel replied:
done

labkey-bpatel replied:
prod build did not like ?. syntax, so this is now changed to:
(d["assayTutorialDocuments"] && d["assayTutorialDocuments"].length > 0) || (d["assayTutorialLinks"] && d["assayTutorialLinks"].length > 0)

XingY commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
+        {
+            if (null == form.getDocumentId())
+                errors.reject(ERROR_MSG, "documentId parameter required");
+            if (null == form.getFilename())
```
Can use StringUtils.isBlank here

labkey-bpatel replied:
done

XingY commented on webapp/Connector/src/app/view/module/AssayTutorial.js:
```diff
@@ -0,0 +1,103 @@
+    plugins : ['documentvalidation'],
+
+    statics : {
+        openPopupModal : function(videoUrl) {
```
Seems this is similar to to the win obj in Started.js maybe can reuse?

labkey-bpatel replied:
i had tried reusing, however it just did not work as expected, so went this route

XingY commented on webapp/Connector/src/app/view/module/AssayTutorial.js:
```diff
@@ -0,0 +1,103 @@
+            '<tpl if="assayTutorialDocuments && assayTutorialDocuments.length &gt; 0">',
+                '<table class="learn-study-info">',
+                    '<tpl for="assayTutorialDocuments">',
+                        '<tr><td>',
```
check isLinkValid?

labkey-bpatel replied:
done

XingY reviewed (commented):
Just a few comments.

XingY commented on test/sampledata/dataspace/cdsimport/Document.txt:
```diff
@@ -175,4 +175,9 @@ document_id	label	filename	document_type	assay_identifier
+303	NAB tutorial A	assay_documents/nab_tutorial_a.pdf	Assay Tutorial Document	NAB A3R5
+304	NAB tutorial B	assay_documents/nab_tutorial_b.docx	Assay Tutorial Document	NAB A3R5
+306	Intracellular Cytokine Staining tutorial	assay_documents/ics_tutorial.pdf	Assay Tutorial Document	ICS
+305	NAB YouTube link 1		Assay Tutorial Link	NAB A3R5	https://www.youtube.com/embed/Q3BPWNYyv14	Introduction to NAb (Neutralizing antibody assay)
```
:thumbsup: good job Chef Kobe

labkey-bpatel replied:
Haha! indeed!

labkey-bpatel committed:
Code review changes:
- change confusing function name
- remove unused variable check
- check for null
- check whether the document link is valid
- check for study public access

labkey-bpatel committed:
Prod build fix - simplified syntax to check for null (?.) is throwing an exception, so changing to basic syntax.
